Ability to withdraw payment cases that are yet to be submitted to service provider

ajeety4 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

Ability to withdraw payment cases that are yet to be submitted to service provider i.e. in the `Pending Submission` state.

Screenshots:
- Button with confirmation modal on click

<img width="2676" height="1060" alt="image" src="https://github.com/user-attachments/assets/e2c27cdc-f88c-4877-9df0-affc31bb0f73" />

- After revert verification
<img width="2619" height="865" alt="image" src="https://github.com/user-attachments/assets/2874c7d8-9cdf-4676-bafd-1feccc8d8354" />

- Reverting verification for non eligible statuses

<img width="2724" height="1110" alt="image" src="https://github.com/user-attachments/assets/78a72c21-e3b8-478a-8372-40b1af7a6a97" />


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
[Ticket](https://dimagi.atlassian.net/browse/SC-4547)

Additionally, 
- it also adds validation in the backend for max no. of cases that can be verified and reverted. The count is set to 100 which is same as the max value of rows that can be selected from the UI.
- Adds validation so that only cases that are `Not Verified` or `Request Failed` are eligible for verification. This avoids any accidental duplicate payments.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
MTN_MOBILE_WORKER_VERIFICATION

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested on local.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
New test cases included.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
None

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

ajeety4 committed:
function to revert verified cases

ajeety4 committed:
add alert in UI for revert verification message

ajeety4 committed:
add view method to revert payment verification

ajeety4 committed:
add button and confirmation modal for revert verification

ajeety4 committed:
css for buttons

ajeety4 committed:
javascript to enable button on selection of rows

ajeety4 committed:
add request validation for no case id and max limit

ajeety4 committed:
add validation for verification request

ajeety4 committed:
add validation of payment status for verification request

ajeety4 committed:
nit: split functions

ajeety4 committed:
isort

ajeety4 committed:
update both buttons on refresh of data

ajeety4 committed:
custom error modal for htmx errors

kaapstorm commented on corehq/apps/integration/payments/services.py:
```diff
@@ -118,7 +127,7 @@ def verify_payment_cases(domain, case_ids: list, verifying_user: WebUser):
     for case_ids_chunk in chunked(case_ids, CHUNK_SIZE):
         cases_updates = _get_cases_updates(case_ids_chunk, payment_properties_update)
-        _, cases = handle_case_update(
+        xform, cases = handle_case_update(
```
No need to change this, but by convention in the codebase we use `__` (double underscore) for unused variables so that it doesn't clash with `_`.

ajeety4 commented on corehq/apps/integration/payments/services.py:
```diff
@@ -197,3 +197,32 @@ def get_payment_batch_numbers_for_domain(domain):
+    for case in CommCareCase.objects.iter_cases(case_ids, domain):
+        payment_status_value = case.get_case_property(PaymentProperties.PAYMENT_STATUS)
+        if PaymentStatus.from_value(payment_status_value) != PaymentStatus.PENDING_SUBMISSION:
+            raise PaymentRequestError(
```
> If you raise exception here, some cases might get updated and the user would not know that?

Raising the exception actually skips all the cases since updates happen separately [below](https://github.com/dimagi/commcare-hq/pull/36741/files#diff-576cb3cd65ff10b9ee679889c40e32831327071a632295781ebd4a77eaf28745R229).

ajeety4 commented on corehq/apps/integration/templates/payments/payments_verify_report.html:
```diff
@@ -36,6 +36,18 @@ <h1 class="py-3 m-0">{% trans "Payments Verification Report" %}</h1>
+      data-bs-toggle="modal"
+      data-bs-target="#revert-verification-confirmation-modal"
+    >
+      <span> {% trans "UnVerify Selected" %} </span>
```
How does `Revert Verification` sounds to you ?

kaapstorm replied:
:+1: Sounds good.

ajeety4 committed:
message update

ajeety4 committed:
nit: unnecessary casting

ajeety4 committed:
nit: reuse

ajeety4 commented on corehq/apps/integration/payments/services.py:
```diff
@@ -197,3 +197,32 @@ def get_payment_batch_numbers_for_domain(domain):
+            )
+
+    payment_properties_update = {
+        PaymentProperties.PAYMENT_VERIFIED: str(False),
```
Addressed in https://github.com/dimagi/commcare-hq/pull/36741/commits/cae71f68c7d46763288988353336f1307ef7ead5

ajeety4 committed:
nit: wordings update

ajeety4 commented on corehq/apps/integration/templates/payments/payments_verify_report.html:
```diff
@@ -36,6 +36,18 @@ <h1 class="py-3 m-0">{% trans "Payments Verification Report" %}</h1>
+      data-bs-toggle="modal"
+      data-bs-target="#revert-verification-confirmation-modal"
+    >
+      <span> {% trans "UnVerify Selected" %} </span>
```
Addressed in https://github.com/dimagi/commcare-hq/pull/36741/commits/60a598adfbc8e8143ee5dff38548de5a2c905019.  Used `Revert Verification`

ajeety4 replied:
Yeah, we are using `Verify` in both KYC and MoMo. I feel `Approve` for payments is more suitable, and `Revert Approval` works better for the reverse process. Let me run this through AE and get their thoughts as well.

We currently don’t have a Reject or Hold Off functionality, but I think it could be valuable to have it in future,

ajeety4 replied:
I checked with Brendon on this and we are good with using `Approve` and `Withdraw Approval` . Additionally, on the same lines, we can rename the report from `Payments Verification Report`  to `Payments Approval Report`. 
I am going to do this in a separate PR as I would like to take care of the naming in the code as well accordingly. (i.e View, table and method names etc. )

mkangia commented on corehq/apps/integration/payments/services.py:
```diff
@@ -197,3 +197,32 @@ def get_payment_batch_numbers_for_domain(domain):
     return [batch_number for batch_number in batch_numbers if batch_number]
+
+
+def revert_payment_verification(domain, case_ids: list):
```
ya, this needs some thinking through.

The approach to be taken would depend upon how we send payments from the task and how we un-verify them here, i.e. in bulk or one by one.

Blocking the un-verify process while the task is running is the simplest way here.

You can also check out `get_redis_lock` function to do object level locking.

ajeety4 committed:
block revert verification if payments submissions are active for the domain

mkangia reviewed (approved):
Looks good, just some non blocking code improvements to consider.

ajeety4 committed:
nit: simplification

ajeety4 committed:
nit: use label instead of string

took oppurtunity to add a test case

ajeety4 committed:
nit: split function

ajeety4 committed:
nit: ordering

ajeety4 committed:
disallow parallel requests for revert verification for a domain

ajeety4 committed:
Additional safeguard to ensure we only process payment cases that are pending submission

ajeety4 committed:
nit: function split

ajeety4 committed:
Implements a one-time retry for domains with active revert verification request

ajeety4 commented on corehq/apps/integration/payments/services.py:
```diff
@@ -197,3 +210,30 @@ def get_payment_batch_numbers_for_domain(domain):
+
+    if not _validate_payment_status_for_cases(case_ids, domain, [PaymentStatus.PENDING_SUBMISSION]):
+        raise PaymentRequestError(
+            _("Only payments in the 'Pending Submission' state are eligible for verification reversal.")
```
a2691ff8ccbb0c385717929e38084f9e3b7651bb

ajeety4 commented on corehq/apps/integration/payments/services.py:
```diff
@@ -118,14 +123,22 @@ def verify_payment_cases(domain, case_ids: list, verifying_user: WebUser):
+        payment_status_value = case.get_case_property(PaymentProperties.PAYMENT_STATUS)
+        if PaymentStatus.from_value(payment_status_value) not in valid_statuses:
+            return False
+    return True
```
Addressed ib fae8190b414a7262f49546c322b93418b64852f1

ajeety4 commented on corehq/apps/integration/payments/views.py:
```diff
@@ -175,16 +193,30 @@ def _apply_filters(self, query):
+
+        if len(case_ids) > self.VERIFICATION_ROWS_LIMIT:
+            raise PaymentRequestError(
+                _("You can only verify for up to {} cases at a time.").format(self.VERIFICATION_ROWS_LIMIT)
```
afaa7351c72fcc31378b703d65ceac4586fe9df1

ajeety4 commented on corehq/apps/integration/payments/views.py:
```diff
@@ -196,6 +228,42 @@ def verify_rows(self, request, *args, **kwargs):
             context,
         )
 
+    def _validate_revert_verification_request(self, case_ids):
```
Addressed in 694591d9781218cada87a039954a7d69f6527e58

ajeety4 commented on corehq/apps/integration/payments/views.py:
```diff
@@ -196,6 +228,42 @@ def verify_rows(self, request, *args, **kwargs):
+
+        try:
+            self._validate_revert_verification_request(case_ids)
+            self._check_for_active_payment_task()
```
Addressed in 694591d9781218cada87a039954a7d69f6527e58

mkangia commented on corehq/apps/integration/payments/services.py:
```diff
@@ -40,6 +40,10 @@ def _get_payment_cases_updates(case_ids_chunk, config):
+        # Additional safeguard to ensure we only process cases that are pending submission
+        payment_status_value = payment_case.get_case_property(PaymentProperties.PAYMENT_STATUS)
+        if PaymentStatus.from_value(payment_status_value) != PaymentStatus.PENDING_SUBMISSION:
+            continue
```
I dropped an approval which is good to go if this is not a concern.

ajeety4 committed:
fix: order of check and memoized

ajeety4 committed:
fix test cases

adds missing payment status, add mock for active revert verification request tracker
